scico/denoiser.py

```diff
@@ -27,6 +34,61 @@
 from ._flax import FlaxMap
 
 
+def bm4d(x: JaxArray, sigma: float):
+    r"""An interface to the BM4D denoiser :cite:`maggioni-2012-nonlocal`.
+
+    BM4D denoising is performed using the
+    `code <https://pypi.org/project/bm4d/>`__ released by the authors of
+    :cite:`maggioni-2012-nonlocal`.
+
+    Args:
+        x: Input image.
+        sigma: Noise parameter.
+
+    Returns:
+        Denoised output.
+    """
+    if not have_bm4d:
+        raise RuntimeError("Package bm4d is required for use of this function.")
+
+    bm4d_eval = tunibm4d.bm4d
+
+    if np.iscomplexobj(x):
+        raise TypeError(f"BM4D requires real-valued inputs, got {x.dtype}")
+
+    # Support arrays with more than three axes when the additional axes are singletons.
+    x_in_shape = x.shape
+
+    if isinstance(x.ndim, tuple) or x.ndim < 3:
+        raise ValueError(f"BM4D requires three dimensional (M, N, P) inputs; got ndim = {x.ndim}")
```

Comment:
Does the `(M, N, P)` without explanation play any useful role here?

Reply:
I essentially extended what bm3d did (stating `(M,N)` and `(M,N,C)' without using it later). Should we remove it for both bm4d and bm3d?

Reply:
True: also not very helpful in `bm3d`. I would remove those from the exception text and add the missing descriptions of expected array shapes to the docs.

Reply:
Fixed it.